plugins/souk21/commandpalette/widgets_commandpalettewidget.js

```diff
@@ -727,14 +728,31 @@ Command Palette Widget
 			}
 			else {
 				searches = this.searchSteps.reduce((a, c) => [...a, ...c(terms)], []);
-				searches = Array.from(new Set(searches));
+				let _unique = new Set();
+				// Filter search results to only get the first unique occurence of every tiddler name
+				searches = searches.filter(s => {
+					if (_unique.has(s.name))
+						return false;
+					_unique.add(s.name);
+					return true;
+				});
 			}
 			this.showResults(searches);
 		}
 
 		searchStepBuilder(filter, caret, hint) {
 			return (terms) => {
-				let search = filter.substr(0, caret) + terms + filter.substr(caret);
+				let search = "";
+				if (caret) {
+					// Use legacy "caret" logic
+					search = filter.substr(0, caret) + terms + filter.substr(caret);
+				}
+				else if (filter.indexOf("regexp[") !== -1 || filter.indexOf("regexp:title[") !== -1) {
+					search = filter.replace("%s", Utils.escapeRegExp(terms));
```

Comment:
If this change is being done  I think it would make more sense to use `/%s/g` to allow multiple occurrences to be replaced rather than just the first one.
Or even better I'd love to see an implementation which just uses TW variables, so the filters would look like `[search<query>]` since that would be the most versatile even if slower. Then you could also have a variable `query-escaped` to use inside `regexp`.

Reply:
You're right about the first part, and I fixed that now.

About using a TW variable instead of using static text, I think it would be more complicated and less flexible. E.g. if my query is now `[(?i)\b%s]`, I cannot simply replace this by `<query>`, as I want the "(?i)\b" prefix.

Reply:
I can't agree it would be more complicated (unless you mean implementation complication) - if anything, for an existing TW user it would make perfect sense since it uses established conventions. Agree with the flexibility though.

Part of the reason I'd like to see an implementation using variables is because that would allow using square brackets in the query without breaking the filter; even if tiddlers with square brackets are not recommended. I just don't like leaving such loose ends unhandled.
Another part is I like writing idiomatic code in my plugins to the best of my abilities.
And now that I think about it, the problem of lack of flexibility could be solved by adding optional key `queryTransformFilter` that would transform the query for a given step.

Though I have to say, in a way it would be mostly art for its own sake.

Reply:
I spent some time trying to do this, and came up with PR #20.